libsolidity/lsp/DocumentHoverHandler.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/lsp/DocumentHoverHandler.h>
+#include <libsolidity/lsp/Utils.h>
+
+#include <fmt/format.h>
+
+namespace solidity::lsp
+{
+
+using namespace std;
+
+using namespace solidity::lsp;
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+namespace
+{
+
+struct MarkdownBuilder
+{
+	std::stringstream result;
+
+	MarkdownBuilder& code(std::string const& _code)
+	{
+		auto constexpr SolidityLanguageId = "solidity";
+		result << "```" << SolidityLanguageId << '\n' << _code << "\n```\n\n";
+		return *this;
+	}
+
+	MarkdownBuilder& paragraph(std::string const& _text)
+	{
+		if (!_text.empty())
+		{
+			result << _text << '\n';
+			if (_text.back() != '\n') // We want double-LF to ensure constructing a paragraph.
+				result << '\n';
+		}
+		return *this;
+	}
+};
+
+}
+
+void DocumentHoverHandler::operator()(MessageID _id, Json::Value const& _args)
+{
+	auto const [sourceUnitName, lineColumn] = HandlerBase(*this).extractSourceUnitNameAndLineColumn(_args);
+	auto const [sourceNode, sourceOffset] = m_server.astNodeAndOffsetAtSourceLocation(sourceUnitName, lineColumn);
+
+	MarkdownBuilder markdown{};
+	//markdown.paragraph(fmt::format("This is: {}", typeid(*sourceNode).name()));
+	auto rangeToHighlight = toRange(sourceNode->location());
+
+	// Try getting the type definition of the underlying AST node, if available.
+	if (auto const* expression = dynamic_cast<Expression const*>(sourceNode))
+	{
+		if (auto const* declaration = ASTNode::referencedDeclaration(*expression))
+			if (declaration->type())
+				markdown.code(declaration->type()->toString(false));
+	}
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(sourceNode))
+	{
+		if (declaration->type())
+			markdown.code(declaration->type()->toString(false));
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(sourceNode))
+	{
+		for (size_t i = 0; i < identifierPath->path().size(); ++i)
+		{
+			if (!identifierPath->pathLocations()[i].containsOffset(sourceOffset))
+				continue;
```

Comment:
You could `break` after finding the first matching segment of the identifier path. They can't overlap, can they?

Reply:
yeah they cannot. improved. :)